### ISSUE TITLE
TimeOfDay subtraction

### ISSUE THREAD

**Issue by mj1856** [mj1856/corefx-dateandtime#33]
This is too wordy:

```csharp
TimeSpan duration = TimeOfDay.CalculateDuration(startTime, endTime);
```

It can be simplified as
```csharp
TimeSpan duration = endTime.Subtract(starttime);
```
and also
```csharp
TimeSpan duration = endTime - starttime;
```

Since subtraction is a [noncommutative operation](https://en.wikipedia.org/wiki/Commutative_property), this works just fine.  However, since we loop over midnight, recognize that while `02:00 - 01:00 == 1 hour`, the inverse would be `01:00 - 02:00 == 23 hours`

**Comment by ashmind** [comment]
> Remember, this is a 24-hour circular clock. There's no associated date. If you pick two points on the clock and ask for the duration between them, it's always going to calculate in a clockwise motion.

I understand the logic as it is here, but I don't quite understand the reasoning. Why can't it calculate assuming it is the same day?

**Comment by ashmind** [comment]
> It's more intuitive if you think about values that cross midnight, such as 02:00 - 22:00 = 4 hours. This happens in real-world scenarios when employees work night shifts.

That's a fair point. Though without date it could just as well be 28h.
Maybe not for a shift of course, but still feels a bit weird.

**Comment by mj1856** [comment]
Lol.  Yes, it could be 28h.  But if we went the other way, it could also be -20h or -44h, etc.  :smile: 

Thus, we need some rules to avoid ambiguity.  The rules I've chosen are to always will move clockwise, and will stop at the first match encountered.  Since time moves forward (for most people), I think this makes sense.  
![image](https://cloud.githubusercontent.com/assets/1396388/8347987/1bf20f24-1ac2-11e5-9895-5f00df2efa11.png)

**Comment by jodastephen** [comment]
JSR-310 and Joda-Time support negative amounts of time and calculate within the bounds of the day. So the result is always -24 hours > result > +24 hours. Given that `TimeSpan` also supports negative values, this is the correct approach IMO.

**Comment by mj1856** [comment]
@jodastephen - to clarify, are you saying that the *current implementation* is correct?

```
1:00 am - 11:00 pm == 2 hours
```

Or are you suggesting that the result should go negative instead?

```
1:00 am - 11:00 pm == -22 hours
```

**Comment by jodastephen** [comment]
Imagine putting the same date in front of each. What answer would you get? (-22 hours)

**Comment by mj1856** [comment]
Sure, but here there is no date... :smile:

Consider usage scenarios.  Example: daily business hours of restaurants, bars, theaters.  If it's a schedule for a particular week, then `DateTime` fine.  But if it's a recurring rule - then the date has to drop off.  Since we don't allow for more than 24 hours in a day (and shouldn't), then one couldn't easily represent a range of values that crosses midnight without this feature.

**Comment by ashmind** [comment]
I think negative should be the _default_, but no reason not to provide a different calculation as well.
E.g. default `-` gives negative, but corresponding `Subtract` method could have a `TimeSubstraction` enum to allow a different option.

**Comment by jodastephen** [comment]
With date and time, it is important to keep each concept separate. The rule in JSR-310/Joda-Time is simple - both times are considered to be attached to the same invisible date, thus negatives result.

Any user that wants behaviour to do with time over midnight needs to manage that themselves, its not up to a core library to make that use case assumption.

FWIW, I think that this operator overload is wrong anyway.

    LocalTime + TimeSpan  = LocalTime
    LocalTime - TimeSpan  = LocalTime

    LocalTime + LocalTime = ???
    LocalTime - LocalTime = TimeSpan ??? not really

Adding/Subtracting a duration to a local-time is fine. But adding two local times is nonsense, thus subtracting one local time from another is also nonsense.

**Comment by mj1856** [comment]
I originally had this as `TimeOfDay.CalculateDuration`.  However I noticed that both `DateTime` and `DateTimeOffset` already have implemented a subtraction operator which returns a `TimeSpan`, without a corresponding addition operator.   Therefore, I consider that an established .NET Framework idiom, and thought we should snap to it.

Consider also that many other methods in this type already have the behavior of crossing over midnight seamlessly.  For example:

```csharp
    new TimeOfDay(23,00).AddHours(2) == new TimeOfDay(01,00)
```

Though in the subtraction case, we're returning a `TimeSpan`, so perhaps that's a key differentiator.

I think perhaps we need both.  The subtraction operator would be implemented to return negative, and the `CalculateDuration` method can be used when crossing over midnight.
